File: 63. Unique Paths II.cpp
#include <bits/stdc++.h>

using namespace std ; 

int uniquePathsWithObstacles(vector<vector<int>>& v, vector<vector<int>>& dp, int i, int j)
{
	if(v[i][j])
		return 0 ; 

	if(i == 0 or j == 0)
		dp[i][j] = 1 ;

	if(dp[i][j] == 0)
	{
		dp[i][j] = uniquePathsWithObstacles(v, dp, i-1, j) + uniquePathsWithObstacles(v, dp, i, j-1) ; 
	} 

	return dp[i][j] ; 
}

int uniquePathsWithObstacles(vector<vector<int>>& v) 
{
	int res = 0 ; 
	vector<vector<int>> dp(v.size(), vector<int>(v[0].size(), 0)); 
 	return uniquePathsWithObstacles(v, dp, dp.size()-1, dp[0].size()-1);     
}

int uniquePathsWithObstacles1(vector<vector<int>>& v)
{
	int n = v.size(), m = v[0].size() ; 
	vector<vector<int>> dp(n, vector<int>(m, 0));

	for(int i = 0 ; i < n ; i++)
	{
		if(v[i][0] == 0)
			dp[i][0] = 1 ; 
		else 
			break ; 
	}

	for(int j = 0 ; j < m ; j++)
	{
		if(v[0][j] == 0)
			dp[0][j] = 1 ; 
		else 
			break ; 
	}

	for(int i = 1 ; i < n ; i++)
	{
		for(int j = 1 ; j < m ; j++)
		{
			if(v[i][j])
				dp[i][j] = 0 ; 
			else
				dp[i][j] = dp[i-1][j] + dp[i][j-1] ; 
		}
	}

	return dp[n-1][m-1] ; 
}

int main()
{
	vector<vector<int>> v = {
		{0}, {0}
	}; 

	cout << uniquePathsWithObstacles1(v) ; 

	return 0 ; 
}

File: Java/63. Unique Paths II.java
import java.util.Arrays;

class Solution {
    /*
     * Brute-Force-Method
     * Time: O(2^(m+n))
     * Space: O(m+n)
     */
    public int uniquePathsWithObstacles_1(int[][] grid) {
        return uniquePathUtil(grid, 0, 0);
    }

    private int uniquePathUtil(int[][] grid, int i, int j) {
        if (i == grid.length || j == grid[0].length || grid[i][j] == 1)
            return 0;

        if (i == grid.length - 1 && j == grid[0].length - 1)
            return 1;

        return uniquePathUtil(grid, i + 1, j) + uniquePathUtil(grid, i, j + 1);
    }

    /*
     * DP-Top-Down
     * Time: O((m*n))
     * Space: O(m*n)
     */
    public int uniquePathsWithObstacles_2(int[][] grid) {
        int[][] cache = new int[grid.length][grid[0].length];

        for (int i = 0; i < grid.length; i++)
            Arrays.fill(cache[i], -1);

        return uniquePathUtil(grid, 0, 0, cache);
    }

    private int uniquePathUtil(int[][] grid, int i, int j, int[][] cache) {
        if (i == grid.length || j == grid[0].length || grid[i][j] == 1)
            return 0;

        if (i == grid.length - 1 && j == grid[0].length - 1)
            return 1;

        if (cache[i][j] == -1)
            cache[i][j] = uniquePathUtil(grid, i + 1, j, cache) + uniquePathUtil(grid, i, j + 1, cache);

        return cache[i][j];
    }

    /*
     * DP-Bottom-Up
     * Time: O((m*n))
     * Space: O(m*n)
     */
    public int uniquePathsWithObstacles(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] cache = new int[m][n];

        // fill the 1st row
        for (int i = 0; i < m; i++) {
            if (grid[i][0] == 0)
                cache[i][0] = 1;
            else
                break;
        }

        // fill the 1st col
        for (int j = 0; j < n; j++) {
            if (grid[0][j] == 0)
                cache[0][j] = 1;
            else
                break;
        }

        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (grid[i][j] == 0)
                    cache[i][j] = cache[i - 1][j] + cache[i][j - 1];
            }
        }

        return cache[m - 1][n - 1];
    }
}
